gpytorch/mlls/leave_one_out_likelihood.py

```diff
@@ -0,0 +1,66 @@
+import math
+
+import torch
+from torch import Tensor
+
+from ..distributions import MultivariateNormal
+from .exact_marginal_log_likelihood import ExactMarginalLogLikelihood
+
+
+class LeaveOneOutLikelihood(ExactMarginalLogLikelihood):
+    """
+    The leave one out cross-validation (LOO-CV) likelihood from RW 5.4.2 for an exact Gaussian process with a
+    Gaussian likelihood. This offers an alternative to the exact marginal log likelihood where we
+    instead maximize the sum of the leave one out log probabilities :math:`\log p(y_i | X, y_{-i}, \theta)`.
+
+    Naively, this will be O(n^4) with Cholesky as we need to compute `n` Cholesky factorizations. Fortunately,
+    given the Cholesky factorization of the full kernel matrix (without any points removed), we can compute
+    both the mean and variance of each removed point via a bordered system formulation making the total
+    complexity O(n^3).
+
+    The LOO-CV approach can be more robust against model mis-specification as it gives an estimate for the
+    (log) predictive probability, whether or not the assumptions of the model is fulfilled.
+
+    .. note::
+        This module will not work with anything other than a :obj:`~gpytorch.likelihoods.GaussianLikelihood`
+        and a :obj:`~gpytorch.models.ExactGP`. It also cannot be used in conjunction with
+        stochastic optimization.
+
+    :param ~gpytorch.likelihoods.GaussianLikelihood likelihood: The Gaussian likelihood for the model
+    :param ~gpytorch.models.ExactGP model: The exact GP model
+
+    Example:
+        >>> # model is a gpytorch.models.ExactGP
+        >>> # likelihood is a gpytorch.likelihoods.Likelihood
+        >>> loocv = gpytorch.mlls.LeaveOneOutLikelihood(likelihood, model)
+        >>>
+        >>> output = model(train_x)
+        >>> loss = -loocv(output, train_y)
+        >>> loss.backward()
+    """
+
+    def __init__(self, likelihood, model):
+        super().__init__(likelihood=likelihood, model=model)
+        self.likelihood = likelihood
```

Comment:
should probably raise an error here if a non-gaussian LH is passed in

Reply:
That should happen in the `super` call since we inherit from `ExactMarginalLogLikelihood`.